pkg/proctree/exit_processing.go

```diff
@@ -0,0 +1,88 @@
+package proctree
+
+import (
+	"github.com/aquasecurity/tracee/signatures/helpers"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+// processExitEvent remove references of processes from the tree when the corresponding process exit without children, or
+// if the last child process of a process exits.
+// Notice that there is a danger of memory leak if there are lost events of sched_process_exit (but is limited to the
```

Comment:
I would say it is quite certain there will be leaks because its quite certain there will be lost events for process creation/exits. 

Could you clarify for me what happens if:

- lost a process exit for a child of a process
- parent exists and isn't cleared because of a fake existing child
- parent PID is re-used by the OS and `processDefaultEvent` would pick the "wrong process from the tree" (from the old existing process, with a different parent and outdated metadata).

I bet this assumption is wrong, maybe I haven't read the code enough, but could you clarify how is this solved ?

Reply:
I bet this is solved at `addNewForkdProcess` (it states that "this will delete old instance if its exit was missing") but I miss how.

1. processForkEvent processes a fork event
2. it calls processDefaultEvent (it will find the old event and add a thread to it, possibly with other existing threads to already exited process)
3. processMainThreadFork or processThreadFork will call getProcess

there is a comment saying that getProcess will have errors if the existing process is a result of lost exit event. I don't see how.

Reply:
The idea is the following:
* We can't know if we missed the exit of a process or one of its children.
* The only way to know this has happened, is if we get an event that tells us that an existing process in the tree was forked (means that its PID didn't exist before, and now the new fork process takes its place)
* So, when we get fork event we check if the process in the process tree is found as forked already.
* If it was forked, we overwrite it with the new fork information.

Now, I see your point here and thought about few things:
* We can also determine that we missed exit of a thread if we see a thread event with different PID. So this should also increase our accuracy in the data we have.
* Currently I only override the current process information, leaving its parents and children with false reference. This is really not good. What we can do here is just call the delete function to remove its information from the tree.

In the end, it is impossible to create a process tree which is safe from lost events, because you can't expect to have an accurate data source created by inaccurate data. We do a best effort here to notice if the information we get does not agree with the information we had until now.

Reply:
> * This is really not good. What we can do here is just call the delete function to remove its information from the tree.

Yep, even if we do more flushes than we wanted, I believe it will be "safer" considering the complexity of the updates and the possibility that we miss events (unfortunately). Its tricky, indeed.

>  We do a best effort here to notice if the information we get does not agree with the information we had until now.

YES, that was my point about being skeptical. Maybe we should do 2 things to make sure we are "good enough":

1. do some full cleanups from time to time (like a garbage collector approach, either cleaning things up if needed OR even flushing lots of references just for the sake of keeping things right -> we need to confirm if this is even needed).
2. a way to confirm this is needed would be to stress test (functional tests, not unit) the tree with lots of process creation/deletion/forks/pthreads creation. then take pictures of the tree and check how bad (or good) it is.

that would give me a good sensation about this, and unfortunately I know its a bit more of work... sorry for asking that.. I just wanted to find a way to be confident that we can trust in the tree enough (after hours/days and not only for 1 or 2 hours).

Reply:
This was solved now using a capped size LRU for garbage collecting nodes.
We can improve it to also limit the amount of dead ancestors we keep for a node, but this could come later on.